fallback/s2a_fallback_test.go

```diff
@@ -0,0 +1,101 @@
+/*
+ *
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package fallback
+
+import (
+	"testing"
+)
+
+func TestDefaultFallbackClientHandshakeFunc(t *testing.T) {
+	for _, tc := range []struct {
+		desc              string
+		inputFallbackAddr string
+		funcInitialized   bool
+		expectNilErr      bool
+	}{
+		{
+			"working case, fallback address has port suffix",
+			"example.com:443",
+			true,
+			true,
+		},
+		{
+			"working case, fallback address has no port suffix",
+			"example.com",
+			true,
+			true,
+		},
+		{
+			"test empty fallback address",
+			"",
+			false,
+			false,
+		},
+	} {
+		fbFunc, err := DefaultFallbackClientHandshakeFunc(tc.inputFallbackAddr)
+		if got, want := fbFunc != nil, tc.funcInitialized; got != want {
```

Comment:
Sorry that I missed this before: in the success cases, I'm concerned that we're not adequately testing fbFunc, since we're just checking that it is non-nil. In the success cases, can we also check that fbFunc behaves as expected e.g. with fake/mock inputs?

Reply:
I think there are ways we can test *parts* of this without relying on a starting up a fake server, but for a complete success test I agree we need e2e tests. For example,

- Factor out the creation of TLS config (which is duplicated code) and unit test this function independently.
- For `DefaultFallbackDialerAndAddress`, we can check that the returned tls.Dialer is exactly the struct we expect.
- Create a non-exported function defaultFallbackClientHandshakeFuncInternal that allows us to inject the functionality in lines 77-78, for example:

```
func DefaultFallbackClientHandshakeFunc(fallbackAddr string) (FallbackClientHandshake, error) ... {
 // call defaultFallbackClientHandshakeFuncInternal and pass in internalDialer
}

func defaultFallbackClientHandshakeFuncInternal(fallbackAddr string, config *tls.Config, dialer func(ctx, string, *tls.config) (net.Conn, error)) (FallbackClientHandshake, error) ... {
  // current implementation but call dialer instead of lines 77-78
}

func internalDialer(ctx context.Context, fallbackAddr string, config *tls.Config) (net.Conn, error) {
  fallbackDialer := &tls.Dialer{Config: config}
  return fallbackDialer.DialContext(ctx, "tcp", fallbackServerAddr)
}
```

WDYT?